pkg/pg/txdb_test.go

```diff
@@ -0,0 +1,69 @@
+package pg
+
+import (
+	"database/sql"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	_ "github.com/marcboeker/go-duckdb"
+
+	"github.com/google/uuid"
+	"github.com/jmoiron/sqlx"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"context"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/utils/tests"
+)
+
+func TestTxDBDriver(t *testing.T) {
+	dbURL, ok := os.LookupEnv("CL_DATABASE_URL")
+	if !ok {
+		fmt.Sprintf("CL_DATABASE_URL not set--falling back to testing txdb backed by an in-memory db")
+		dbURL = string(InMemoryPostgres)
+	}
+	db := NewSqlxDB(t, dbURL)
```

Comment:
Would it make sense to package this up as a test helper?

Reply:
We could, although I can't think of a good use case for this other than testing txdb itself (ie, this test). So making it a test helper, especially if it's in a place where it can be imported, might confuse someone.

Tests that depend on the db migration scripts being run (ie, require the standard tables and indexes set up) should call `NewSqlxDB()` to connect to an actual db and just fail if they can't.  Tests that don't depend on those scripts should call `NewInMemoryDataSource()`. Usually if testing with `NewInMemoryDataSource` is something that can work, then you wouldn't have any reason to attempt testing it with an actual db. In this case, it makes sense because there is some code in the `Open()` method of the txdb driver which parses the db url. This gets skipped if you're testing with an in-memory data source but everything else is tested.  So the full db tests everything but using the in-memory db in its place tests like 95%

Reply:
> Tests that depend on the db migration scripts being run

This is part of the pattern that we are trying to get away from though. It should not be the default.

>  Usually if testing with NewInMemoryDataSource is something that can work, then you wouldn't have any reason to attempt testing it with an actual db

Why do you say this? If we have the ability to trivially integration test against real postgres, then we should use it.

Reply:
> Why do you say this? If we have the ability to trivially integration test against real postgres, then we should use it.

I was mostly saying that because it seems unlikely anything could pass in the in-memory db but fail in a real db. (Unless we're testing the connection to the db itself, or that the db is set up properly... in which case I wouldn't expect an in-memory db to be useful.)  I suppose there are probably some ways, but in general I would assume that the set of sql commands supported by an in-memory db would be a small subset of those supported by a fully fledged database.

> This is part of the pattern that we are trying to get away from though. It should not be the default.

I'm not sure what the motivation for wanting to change existing patterns is, but if the reason is because we want to minimize use of the external db to conserve resources then making this a helper function and encouraging its use would be diametrically opposed to that. If we always try to hit the external db first, and only use the in-memory one as a backup if that fails then we'd waste a whole lot of resources in CI since 100% of them will go to the db. I'm not sure what the purpose of an in-memory db would even be then. In the short term, it could help us with local testing before we get the db isolation work Keith planned out done. But once each relay has its own migration scripts, it shouldn't be a problem.

Reply:
To clarify:

In my last paragraph I was assuming this meant that you think we should always call this helper function instead of using an in-memory db by itself:
>  If we have the ability to trivially integration test against real postgres, then we should use it.

So that's why I say "100% of them will go to the db". If instead by that, you just mean that we ought to leave the option open to do that for some tests, rather than an all-or-nothing approach then my objection above wouldn't apply.

Reply:
Update: in light of having read your other thread, I understand now that the motivation is not conserving CI resources but making it easier for devs to run tests locally. At least for some tests, it may still not be worth using the extra CI resources just to get that slight bit of extra certainty that the function you're testing behaves as expected. But given your motivations, it does seem like you're recommending an all-or-nothing approach. Since unless all of the tests were able to run in-memory you wouldn't get rid of the dependency... right?

Reply:
No, it does not imply that we need to use short at all:
https://github.com/smartcontractkit/chainlink-common/pull/910#discussion_r1859423319

Reply:
Ah, I guess this is what you had in mind...
> If there are dependencies required, then typically you opt-in to those tests by e.g. including a build tag, setting a flag/env-var, etc.

So you're just recommending we move to using one of these methods instead of --short... that makes sense